pickhardtpayments/OracleLightningNetwork.py

```diff
@@ -15,41 +17,82 @@ def __init__(self, channel_graph: ChannelGraph):
+    def allocate_amount_as_inflight_on_path(self, attempt: Attempt):
         """
+        allocates `amt` as in_flights to all channels of the path
+        """
+        for uncertainty_channel in attempt.path:
+            oracle_channel = self.get_channel(uncertainty_channel.src, uncertainty_channel.dest,
+                                              uncertainty_channel.short_channel_id)
+            oracle_channel.in_flight += attempt.amount
 
-        :rtype: object
+    def send_onion(self, attempt: Attempt):
+        """
+        Probes the oracle network if the amount of satoshis for this attempt can be sent through the given path in
+        the attempt.
+        If successful, then inflight amounts are placed on the respective oracle channels as well as uncertainty
+        channels, and the Attempt is set to INFLIGHT.
+        If not successful, then the status of the Attempt is set to FAILED and the failing channel is returned
+
+        :param attempt: the attempt that is probed
+        :type: Attempt
+
+        :return: did sending the onion succeed?
+        :rtype: Boolean
+        :return: if sending the onion failed, at which channel did it fail
+        :rtype: UncertaintyChannel or None
         """
-        for channel in path:
-            oracle_channel = self.get_channel(
-                channel.src, channel.dest, channel.short_channel_id)
-            success_of_probe = oracle_channel.can_forward(
-                channel.in_flight + amt)
-            # print(channel,amt,success_of_probe)
-            channel.update_knowledge(amt, success_of_probe)
+        for uncertainty_channel in attempt.path:
+            oracle_channel = self.get_channel(uncertainty_channel.src, uncertainty_channel.dest,
+                                              uncertainty_channel.short_channel_id)
+            # probing for current amount in addition to current in_flights in oracle network
+            success_of_probe = oracle_channel.can_forward(oracle_channel.in_flight + attempt.amount)
+            if success_of_probe:
+                # replicate HTLCs on the OracleChannels and place inflights in Uncertainty Network
+                logging.debug("allocating {:,} on channel".format(attempt.amount))
+                oracle_channel.in_flight += attempt.amount
+                uncertainty_channel.allocate_inflights(attempt.amount)
             if not success_of_probe:
-                return False, channel
+                logging.debug("{:,} sats failed on channel {}-{} with actual liquidity of {:,} sats (cap: {:,})".
+                             format((oracle_channel.in_flight + attempt.amount), oracle_channel.src[0:4],
+                                    oracle_channel.dest[0:4], oracle_channel.actual_liquidity, oracle_channel.capacity))
+                attempt.status = AttemptStatus.FAILED
+                logging.debug(f"Attempt status: {attempt}")
+                return False, uncertainty_channel
```

Comment:
not important for this patch but generally we could additionally return the index of the channel (e.g. the position of the hop within the attempt / path)